### PR TITLE
Write energy and coordinates to file to track force field minimization

### PR DESCRIPTION
**Pull request by rachelnwalker**
I made these changes to track the coordinates and energy of each atom in a molecule in each step of force field minimization. I found it useful in analyzing/viewing the minimization process. I'm not sure if we would actually want to merge this, but I wanted to make this PR just in case someone else would find this useful. If we do want to keep this, let me know what sort of `#ifdef` I should use.

Here is an animation of the minimization process I made with this data:

https://user-images.githubusercontent.com/39069546/116461830-95762d00-a81d-11eb-97b7-9907d7828bba.mov

### CODE REVIEW

**Comment by ZontaNicola** [comment]
this is super interesting, thank you!
this also looks like the dof step is not doing its job, most of these clashes could have been prevented by flipping some of these side chains on the other side.
Also I can see the energy does seem to go uphill and end up in a conformation that has a higher energy than some pretty early ones... I kind of like the final one better though

File: CoordgenMinimizer.cpp [hunk 1]
```diff
@@ -72,7 +74,18 @@ void CoordgenMinimizer::run()
     }
 
     for (int iterations = 0; iterations < m_maxIterations; ++iterations) {
-        scoreInteractions();
+        float energy = scoreInteractions();
+
+        // #ifdef ...
```
Comment (ZontaNicola) [inline CoordgenMinimizer.cpp:79]: I'll let Dan decide a name for the macro, I'm just wondering if we want that or an environment variable

File: sketcherMinimizer.h [hunk 2]
```diff
@@ -109,6 +109,7 @@ class EXPORT_COORDGEN sketcherMinimizer
     static void canonicalOrdering(sketcherMinimizerMolecule* minMol);
 
     // void initializeFromMolecule(ChmMol& mol);
+    void writeMinimizationData(sketcherMinimizerPointF center, float s, float c);
```
Comment (d-b-w) [inline sketcherMinimizer.h:112]: What do `s` and `c` mean?

File: sketcherMinimizer.cpp [hunk 3]
```diff
@@ -1003,10 +1004,44 @@ void sketcherMinimizer::bestRotation()
...
+            v.rotate(s, c);
+            sketcherMinimizerPointF new_coord = center + v;
+            energy_file << new_coord.x() << "," << new_coord.y() << ";";
+        }
+        energy_file << "\n";
+    }
+    energy_file.close();
+
+    // print atom mapping to output file
+    std::ofstream atom_mapping_file("atom_mapping.txt");
```
Comment (d-b-w) [inline sketcherMinimizer.cpp:1031]: ah, two files, because of the mapping.

File: CoordgenMinimizer.cpp (hunk 1 above)
Comment (d-b-w) [inline CoordgenMinimizer.cpp:79]: Let's call it DEBUG_MINIMIZATION_COORDINATES and have it cover all of this. You can enable defines in the cmake command `-DDEBUG_MINIMIZATION_COORDINATES=ON`.

File: sketcherMinimizer.h (hunk 2 above)
Comment (rachelnwalker) [reply sketcherMinimizer.h:112]: Good question, I think they are the x and y values that define the rotation of the coordinates in 2D space. They are calculated on line 993 of `sketcherMinimizer.cpp`.

When I first tried writing coordinates to a file, I did not apply this rotation and was very confused why the coordinates I printed to a file were rotated versions of the actual coordinates -- turns out we have to perform the rotation :)

Comment (ZontaNicola) [reply sketcherMinimizer.h:112]: sine and cosine of the rotation angle :)

File: CoordgenMinimizer.cpp (hunk 1 above)
Comment (rachelnwalker) [reply CoordgenMinimizer.cpp]: I tried adding this but the file writing doesn't occur after running this command and rebuilding. I just updated this PR -- do you know what I'm missing?

Comment (d-b-w) [reply CoordgenMinimizer.cpp]: Like, you reran cmake and make install? huh. 

Did the `make install` step build anything? I wonder if you need a rebuild after updating the macro

Comment (rachelnwalker) [reply CoordgenMinimizer.cpp]: Yeah, that wasn't working either, I even tried a clean build

**Comment by rachelnwalker** [comment]
`minimization_data.txt` now includes coordinates from the `avoidClashes` step of coordinate generation. Here is an example animation where force field minimization is not required:

https://user-images.githubusercontent.com/39069546/117040571-e0c98900-acbe-11eb-9016-64d9d9e6b541.mp4

File: CoordgenMinimizer.cpp [hunk 4]
```diff
@@ -71,8 +73,25 @@ void CoordgenMinimizer::run()
...
+#ifdef DEBUG_MINIMIZATION_COORDINATES
+        // store data from this minimization step to be written to a file later
+        energy_list.push_back(energy);
+        std::vector<sketcherMinimizerPointF> these_coordinates;
+        for (auto atom : _atoms) {
+            these_coordinates.push_back(atom->coordinates);
+        }
+        all_coordinates.push_back(these_coordinates);
+#endif
+        energy += 1; // avoid build failure -- will be ok after pushing other PR
```
Comment (rachelnwalker) [inline CoordgenMinimizer.cpp:94]: I'll rebase/fix this once https://github.com/schrodinger/coordgenlibs/pull/95 is merged

**Pull request merged**